File: Week11_BFS/max-depth-bfs.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        /* If root node undefined, just return 0 */
        if (root == null)
            return 0;
        int maxDepth = 0;
        Queue<TreeNode> queue = new LinkedList<>();
        /* Root as the first element */
        queue.offer(root);

        while (!(queue.isEmpty())) {
            /* Increase level by 1 when new level traversed */
            maxDepth++;
            /* Find number of nodes in that level */
            int levelNodes = queue.size();
            /* Dequeue nodes equivalent to number of nodes in level */
            for (int i = 0; i < levelNodes; i++) {
                TreeNode node = queue.poll();
                /* Append child nodes if defined */
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
        }
        return maxDepth;
    }
}

File: Week11_BFS/max-depth-bfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # If root node undefined, just return 0
        if not root:
            return 0
        
        queue = deque()
        # root as the first element
        queue.append(root)
        depth = 0
        
        while len(queue) > 0:
            # Increase level by 1 when new level is traversed
            depth += 1
            # Find number of nodes in that level
            levelNodes = len(queue)
            # Dequeue nodes equivalent to number of nodes in level
            for _ in range(levelNodes):
                # dequeue node
                node = queue.popleft()
                # Append child nodes if defined
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        return depth